File: src/minima_harness/tui/tips.py
from __future__ import annotations

import json

from minima_harness.tui.customize import GLOBAL_DIR

# Persisted rotation cursor: each launch / turn / `/tip` advances it so the user sees a *new* tip
# in order rather than the same one (or a random repeat). Lives next to config.env.
STATE_FILE = GLOBAL_DIR / "tips_state.json"

# Curated tips, each spotlighting one distinctive Minima command — the things a new user would
# never find on their own. Every entry leads with the `/command` so the takeaway is the command.
# Plain strings (no Textual import) so this module unit-tests in isolation.
TIPS: tuple[str, ...] = (
    "/recall pulls lessons from past sessions, even across projects",
    "/escalate auto-retries on a stronger model when answer quality looks thin",
    "/confirm shows the routing trade-off panel before each turn so you pick the model",
    "/model auto lets Minima cost-route; /model <id> pins one",
    "/cache reuses a near-duplicate answer for $0 — toggle the semantic cache",
    "/optimize trims your system prompt via Mubit to save tokens",
    "/ledger set <goal> tracks spend against a budget across the whole goal",
    "/judge toggles LLM quality judging of each answer",
    "/edits forces a diff review before any file write",
    "/tree, /fork and /clone branch and revisit your session history",
    "/thoughts streams the model's reasoning live",
    "/cost opens the meter: estimated vs. actual spend per turn",
    "/compact summarizes older context when the window fills up",
    "/prompt inspects the layered Mubit + local system prompt",
    "/stats shows analytics across your last 10 sessions",
)


def format_tip(body: str) -> str:
    """Prefix a tip body with the lightbulb glyph used across the welcome splash and spinner."""
    return f"💡 {body}"


def pick(index: int) -> str:
    """The tip at ``index``, wrapping around the curated list."""
    return TIPS[index % len(TIPS)]


def _read_index() -> int:
    try:
        return int(json.loads(STATE_FILE.read_text()).get("index", 0))
    except Exception:  # noqa: BLE001 - missing/corrupt/unreadable state → start from 0
        return 0


def advance() -> int:
    """Return the next rotation index and persist it (mod the list length).

    Best-effort: an unwritable HOME must never crash the app, so a failed write just means the
    next launch re-reads the old value — the cycle still rotates within the running session.
    """
    nxt = (_read_index() + 1) % len(TIPS)
    try:
        GLOBAL_DIR.mkdir(parents=True, exist_ok=True)
        STATE_FILE.write_text(json.dumps({"index": nxt}))
    except Exception:  # noqa: BLE001 - read-only HOME / no perms: fall back to in-memory value
        pass
    return nxt

File: packages/tui/src/tui/tips.ts
/**
 * Curated command tips — port of minima_harness/tui/tips.py.
 *
 * Each tip spotlights one distinctive Minima command a new user would never find
 * on their own. `/tip` shows the next one in order; a rotation cursor persists to
 * ~/.minima-harness/tips_state.json so each invocation surfaces a fresh tip rather
 * than a repeat.
 */

import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join, resolve } from "node:path";

// Each entry leads with the `/command` so the takeaway is the command itself.
export const TIPS: readonly string[] = [
  "/recall pulls lessons from past sessions, even across projects",
  "/model auto lets Minima cost-route; /model <id> pins one",
  "/plan toggles read-only plan mode before you let it edit",
  "/budget set <usd> tracks spend against a session budget",
  "/judge toggles LLM quality judging of each answer",
  "/thoughts streams the model's reasoning live",
  "/cost opens the meter: estimated vs. actual spend per turn",
  "/compact summarizes older context when the window fills up",
  "/prompt inspects the layered Mubit + local system prompt",
  "/tree, /fork and /clone branch and revisit your session history",
  "/mouse toggles wheel-scroll vs. native terminal select/copy",
  "web_search + web_fetch give the model live web access (set EXA_API_KEY)",
  "apply_patch edits many files atomically in one call",
  "/resume reopens a past session or run by id",
  "/config manages your API keys (MUBIT, ANTHROPIC, EXA, …)",
];

// Rotation cursor lives next to the harness config. Overridable for tests.
let stateDir = resolve(homedir(), ".minima-harness");
export function setTipsStateDir(dir: string): void {
  stateDir = dir;
}
const stateFile = () => join(stateDir, "tips_state.json");

/** Prefix a tip body with the lightbulb glyph. */
export function formatTip(body: string): string {
  return `💡 ${body}`;
}

/** The tip at `index`, wrapping around the curated list. */
export function pick(index: number): string {
  const n = TIPS.length;
  return TIPS[((index % n) + n) % n] as string;
}

/** The index after `index`, wrapping around (pure). */
export function nextIndex(index: number): number {
  return (index + 1) % TIPS.length;
}

function readIndex(): number {
  try {
    const raw = JSON.parse(readFileSync(stateFile(), "utf8")) as { index?: unknown };
    const i = Number(raw.index);
    return Number.isFinite(i) ? i : 0;
  } catch {
    // missing / corrupt / unreadable state → start from 0
    return 0;
  }
}

/**
 * Advance the rotation cursor, persist it (best-effort), and return the tip at the
 * new index. An unwritable HOME never crashes the app — it just means the next
 * launch re-reads the old value.
 */
export function advance(): string {
  const nxt = nextIndex(readIndex());
  try {
    mkdirSync(stateDir, { recursive: true });
    writeFileSync(stateFile(), JSON.stringify({ index: nxt }));
  } catch {
    // read-only HOME / no perms: fall back to the in-memory rotation
  }
  return pick(nxt);
}
